### PR TITLE
CNV-27866: [Doc] expose the downwardMetrics through a virtio-serial channel

### PR DESCRIPTION
**Pull request by aspauldi**
Version(s): 4.16

Issue: [CNV-27866](https://issues.redhat.com/browse/CNV-27866)

Links to docs previews: 

[Exposing downward metrics for virtual machines](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html)

[Enabling or disabling the downwardMetrics feature gate](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-enabling-disabling-feature-gate)

- [Enabling or disabling the downward metrics feature gate in a YAML file](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-enabling-disabling-downward-metrics-feature-gate-yaml_virt-exposing-downward-metrics)

- [Enabling or disabling the downward metrics feature gate from the command line](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-enabling-disabling-downward-metrics-feature-gate-yaml_virt-exposing-downward-metrics)

 [Configuring a downward metrics device](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-configuring-downward-metrics_virt-using-downward-metrics)

[Viewing downward metrics](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-viewing-downward-metrics_virt-exposing-downward-metrics)

-  [ Viewing downward metrics by using the command line](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-viewing-downward-metrics-cli_virt-exposing-downward-metrics)

-  [Viewing downward metrics by using the vm-metrics-tool](https://74920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/monitoring/virt-exposing-downward-metrics.html#virt-viewing-downward-metrics-tool_virt-exposing-downward-metrics)

QE review:
- [x] QE has approved this change.

### CODE REVIEW

File: modules/virt-configuring-downward-metrics.adoc [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-configuring-downward-metrics_virt-using-downward-metrics"]
+= Configuring downward metrics
+
+You enable the capturing of downward metrics for a host VM by creating a configuration file that includes the following options:
+
+* A `downwardMetrics` device
+* A `virtio-serial` port, through which the metrics are exposed
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:12]: This is not required. Just the `downwardMetrics` device, which automatically creates the virtio-serial port.

File: modules/virt-configuring-downward-metrics.adoc [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+You must also enable a `DownwardMetrics` feature gate.
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:14]:
```suggestion
You must also enable a `downwardMetrics` feature gate.
```

File: modules/virt-configuring-downward-metrics.adoc [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+.Prerequisites
+* Administrator privileges are required to enable the feature gate.
+
+.Procedure
+
+. Edit or create a YAML file to expose the metrics by including a `downwardMetrics` device, a `virtio-serial` port, and the `DownwardMetrics` feature gate as shown in the following example:
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:21]:
```suggestion
. Edit or create a YAML file to expose the metrics by including a `downwardMetrics` device as shown in the following example:
```

File: modules/virt-configuring-downward-metrics.adoc [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+* A `downwardMetrics` device
+* A `virtio-serial` port, through which the metrics are exposed
+
+You must also enable a `DownwardMetrics` feature gate.
+
+.Prerequisites
+* Administrator privileges are required to enable the feature gate.
+
+.Procedure
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:19]: I would suggest to first enable the `donwardMetrics` feature gate with:

```bash
$ oc patch hco kubevirt-hyperconverged -n openshift-cnv \
  --type json -p '[{"op": "replace", "path": \
  "/spec/featureGates/downwardMetrics" \
  "value": true}]'
```

File: modules/virt-configuring-downward-metrics.adoc [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    special: vmi-fedora
+  name: vmi-fedora
+spec:
+  domain:
+    devices:
+      downwardMetrics: {} <1>
+      disks:
+      - disk:
+          bus: virtio <2>
+        name: containerdisk
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:39]: This makes reference to the `containerdisk`, which is used for storing the VM boot image. This is not related with `downwardMetrics`.

File: modules/virt-configuring-downward-metrics.adoc [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        memory: 1024M
+  terminationGracePeriodSeconds: 0
+  volumes:
+  - name: containerdisk
+    containerDisk:
+      image: quay.io/containerdisks/fedora:latest
+  configuration:
+    developerConfiguration:
+      featureGates:
+        - DownwardMetrics <3>
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:53]:
```suggestion
```
Please drop.

File: modules/virt-configuring-downward-metrics.adoc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+----
+<1> The `downwardMetrics` device.
+<2> The `virtio-serial` port.
+<3> The `DownwardMetrics` feature gate.
```
Comment (jcanocan) [inline modules/virt-configuring-downward-metrics.adoc:57]:
```suggestion
```
Just the `downwardMetrics` device should be required at this point if the FG is already enabled.

File: virt/monitoring/virt-exposing-downward-metrics.adoc [hunk 8]
```diff
@@ -0,0 +1,17 @@
+:_mod-docs-content-type: ASSEMBLY
+[id="virt-exposing-downward-metrics"]
+= Exposing downward metrics for virtual machines
+include::_attributes/common-attributes.adoc[]
+:context: virt-exposing-downward-metrics
+
+toc::[]
+
+As an administrator, you can expose a limited set of host and virtual machine (VM) metrics to a guest VM by configuring a YAML file. The YAML file must include a `virtio-serial` port, a `downwardMetrics` device, and a `DownwardMetrics` feature gate. Enabling the feature gate helps to ensure additional protection in restricting viewing of the metrics results.
```
Comment (jcanocan) [inline virt/monitoring/virt-exposing-downward-metrics.adoc:9]:
```suggestion
As an administrator, you can expose a limited set of host and virtual machine (VM) metrics to a guest VM by configuring a YAML file. The YAML file must include a `downwardMetrics` device, and the `DownwardMetrics` feature gate has to be enabled. Enabling the feature gate helps to ensure additional protection in restricting viewing of the metrics results.
```

The sentence `Enabling the feature gate helps to ensure additional protection in restricting viewing of the metrics results.`, IMHO, may lead to users to believe that if the feature gate is enabled, a restricted view is presented, which is not true. If the feature gate is not enabled, the user will not able to get any metric (in fact, the user will not be able to create a VM requesting downwardMetrics at all). If enabled, the user will be able to see all downwardMetrics exposed.

The idea behind the feature gate is to help the admin to restrict if VMs will be able or not to pull the downwardMetrics. Therfore, I will rephrase as: 

`Disabling the feature gate helps to ensure additional protection in restricting viewing of the metrics results.`

**Review by jcanocan — COMMENTED** [review]
Nice job! I've added some comments and clarifications. Hope you find them useful :)

File: modules/virt-configuring-downward-metrics.adoc [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+.Procedure
+
+. Edit or create a YAML file to expose the metrics by including a `downwardMetrics` device, a `virtio-serial` port, and the `DownwardMetrics` feature gate as shown in the following example:
++
+.Example downwardMetrics configuration file
+[source,yaml]
+----
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
```
Comment (dominikholler) [inline modules/virt-configuring-downward-metrics.adoc:27]: @jcanocan Can you please provide an example VM, no VMI, using instancetypes?

**Review by dominikholler — COMMENTED** [review]
> * What metrics are collected and should those be mentioned?

I like the example, because it shows that the hostname is included.
I would not add any additional explanation.

@newkit Can you please also have a look at this PR?

File: modules/virt-configuring-downward-metrics.adoc (hunk 9 above)
Comment (jcanocan) [reply modules/virt-configuring-downward-metrics.adoc:27]: Sure! This is a complete and working VM definition using instancetypes:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: fedora
  namespace: default
spec:
  dataVolumeTemplates:
    - metadata:
        name: fedora-volume
      spec:
        sourceRef:
          kind: DataSource
          name: fedora
          namespace: openshift-virtualization-os-images
        storage:
          resources: {}
          storageClassName: hostpath-csi-basic
  instancetype:
    name: u1.medium
  preference:
    name: fedora
  running: true
  template:
    metadata:
      labels:
        app.kubernetes.io/name: headless
    spec:
      domain:
        devices:
          downwardMetrics: {}
      subdomain: headless
      volumes:
        - dataVolume:
            name: fedora-volume
          name: rootdisk
        - cloudInitNoCloud:
            userData: |
              #cloud-config
              chpasswd:
                expire: false
              password: nye8-9j68-qr57
              user: fedora
          name: cloudinitdisk
```

**Comment by aspauldi** [comment]
@jcanocan, Thank you SO much for the information you've provided! It really is wonderful (even the details that won't go into the user doc but are very helpful to understand in general).

**Comment by aspauldi** [comment]
@dominikholler, thanks for your review and for asking for the example VM definition. So very helpful!

**Comment by jcanocan** [comment]
> @jcanocan, Thank you SO much for the information you've provided! It really is wonderful (even the details that won't go into the user doc but are very helpful to understand in general).

Glad to be helpful! :blush:

File: modules/virt-viewing-downward-metrics-tool.adoc [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+= Viewing downward metrics by using the vm-dump-metrics tool
+
+You can use the `vm-dump-metrics` tool to view downward metrics.
+
+.Procedure
+
+* To view the downward metrics from the guest virtual machine, edit or create a YAML file as shown in the following example:
+
+.Example vm-dump-metrics tool configuration file
+[source,yaml]
```
Comment (jcanocan) [inline modules/virt-viewing-downward-metrics-tool.adoc:16]:
```suggestion
[source,xml]
```

File: modules/virt-viewing-downward-metrics-tool.adoc [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-viewing-downward-metrics-tool_{context}"]
+= Viewing downward metrics by using the vm-dump-metrics tool
+
+You can use the `vm-dump-metrics` tool to view downward metrics.
+
+.Procedure
+
+* To view the downward metrics from the guest virtual machine, edit or create a YAML file as shown in the following example:
```
Comment (jcanocan) [inline modules/virt-viewing-downward-metrics-tool.adoc:13]: IMHO this may be confusing for the final user. What users should address are two bash commands (`dnf install ...` and `vm-dump-metrics`), instead of a yaml file. I would reword this as:

`To view the downward metrics from the guest virtual machine, address the following commands:`. 
wdyt?

File: virt/monitoring/virt-exposing-downward-metrics.adoc [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+[id="virt-exposing-downward-metrics"]
+= Exposing downward metrics for virtual machines
+include::_attributes/common-attributes.adoc[]
+:context: virt-exposing-downward-metrics
+
+toc::[]
+
+As an administrator, you can expose a limited set of host and virtual machine (VM) metrics to a guest VM by:
+
+. Enabling a `downwardMetrics` feature gate. Enabling the feature gate helps to ensure additional protection in restricting viewing of the metrics results
```
Comment (jcanocan) [inline virt/monitoring/virt-exposing-downward-metrics.adoc:11]:
```suggestion
. Enabling a `downwardMetrics` feature gate. Disabling the feature gate helps to ensure additional protection in restricting viewing of the metrics results
```

**Comment by aspauldi** [comment]
Hi @jcanocan, Thanks for the additional comments. I also talked with Pan Ousley, one of the CNV writers, and got some good suggestions. Can you take another look and let me know if you can approve the content?

File: modules/virt-viewing-downward-metrics-tool.adoc [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+.Procedure
+
+. Install the `vm-dump-metrics` tool by running the following command:
++
+[source,terminal]
+----
+$ sudo dnf install -y vm-dump-metrics
+----
+
+. Expose the metrics results by running the following command:
```
Comment (jcanocan) [inline modules/virt-viewing-downward-metrics-tool.adoc:20]:
```suggestion
. Fetch the metrics results by running the following command:
```

**Review by jcanocan — COMMENTED** [review]
Nice job! Just a small comment

**Comment by jcanocan** [comment]
> Hi @jcanocan, Thanks for the additional comments. I also talked with Pan Ousley, one of the CNV writers, and got some  good suggestions. Can you take another look and let me know if you can approve the content?

Glad to help here :blush: The content looks good to me, just added a small comment.

> @jcanocan One more question: Is including sudo necessary for the commands? Thanks!

Yes, it is required.

**Comment by aspauldi** [comment]
@jcanocan  We don't typically use 'fetch' in our user docs. How about 'retrieve'?

**Comment by jcanocan** [comment]
> @jcanocan We don't typically use 'fetch' in our user docs. How about 'retrieve'?

Sounds good! IMHO, "Expose" may lead to think that the information is shown outside the VM, which is not true.

**Comment by jcanocan** [comment]
Great work @aspauldi! Thanks a lot!
/lgtm

**Comment by aspauldi** [comment]
Hi @geetikakay, Can you please review this PR? Thanks!

File: modules/virt-configuring-downward-metrics.adoc (hunk 4 above)
Comment (geetikakay) [reply modules/virt-configuring-downward-metrics.adoc:15]: yes looks better but i see there is a standard way followed for all featuregate. 
https://docs.openshift.com/container-platform/4.15/nodes/clusters/nodes-cluster-enabling-features.html

File: modules/virt-configuring-downward-metrics.adoc [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+      volumes:
+        - dataVolume:
+            name: fedora-volume
+          name: rootdisk
+        - cloudInitNoCloud:
+            userData: |
+              #cloud-config
+              chpasswd:
+                expire: false
+              password: nye8-9j68-qr57
```
Comment (geetikakay) [inline modules/virt-configuring-downward-metrics.adoc:62]: password: '<password>'
probably use same format as mentioned : https://docs.openshift.com/container-platform/4.15/virt/virtual_machines/creating_vms_rh/virt-creating-vms-from-cli.html

**Comment by aspauldi** [comment]
@jcanocan , Can you please look at Geetika's comments? It seems to me that enabling this feature gate is done differently from the example/content Geetika points to in the overall OCP docs.

File: modules/virt-configuring-downward-metrics.adoc (hunk 4 above)
Comment (geetikakay) [reply modules/virt-configuring-downward-metrics.adoc:15]: I would suggest using this way : https://docs.openshift.com/container-platform/4.15/virt/vm_networking/virt-accessing-vm-secondary-network-fqdn.html . That's how we do it for deployKubeSecondaryDNS Feature Gate

Comment (jcanocan) [reply modules/virt-configuring-downward-metrics.adoc:15]: I'm not able to enable the `downwardMetrics` feature gate using either this method: https://docs.openshift.com/container-platform/4.15/nodes/clusters/nodes-cluster-enabling-features.html#nodes-cluster-enabling-features-console_nodes-cluster-enabling or this one: https://docs.openshift.com/container-platform/4.15/nodes/clusters/nodes-cluster-enabling-features.html#nodes-cluster-enabling-features-cli_nodes-cluster-enabling
As far as I was able to understand, both methods seem to be more intended for enabling more generic OpenShift features.

File: modules/virt-viewing-downward-metrics-cli.adoc [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+You can view downward metrics by entering a command from inside a guest virtual machine (VM).
+
+.Procedure
+
+* Run the following command:
++
+[source,terminal]
+----
+$ $ sudo sh -c 'printf "GET /metrics/XML\n\n" > /dev/virtio-ports/org.github.vhostmd.1'
```
Comment (jcanocan) [inline modules/virt-viewing-downward-metrics-cli.adoc:17]:
```suggestion
$ sudo sh -c 'printf "GET /metrics/XML\n\n" > /dev/virtio-ports/org.github.vhostmd.1'
```

File: modules/virt-configuring-downward-metrics.adoc (hunk 4 above)
Comment (aspauldi) [reply modules/virt-configuring-downward-metrics.adoc:15]: I'm confused now. 

@jcanocan , are you suggesting that we remove the **Enabling the downward metrics feature gate** and **Disabling the downward metrics feature gate** topics/commands and show the feature gate configuration through a YAML file instead? I thought you said that wouldn't work.

Also, we say nothing earlier about the HyperConverged CR at all, so mentioning it here seems confusing.

Comment (jcanocan) [reply modules/virt-configuring-downward-metrics.adoc:15]: Sorry about the confusion. 

I'm not sure which option would be a better fit for this case.

I can find examples of toggling the feature gates using a [patch](https://github.com/openshift/openshift-docs/pull/74920#discussion_r1587242288) in the current documentation https://docs.openshift.com/container-platform/4.15/virt/storage/virt-automatic-bootsource-updates.html as well as editing the CR directly in an editor.

This leads me to think that there is not a standard way to enable and disable feature gates. IMHO, the [patch](https://github.com/openshift/openshift-docs/pull/74920#discussion_r1587242288) approach is more straight away and less error prune. WDYT @geetikakay?

In any case, maybe we can squash both sections into one (e.g., Managing downward metrics feature gate), indicating that a `true` value on the `spec.featureGates.downwardMetrics` will enable the FG while a `false` will disable it. Another option is to also squash both sections into one, as previously suggested, and show both patches commands, but following this idea: https://docs.openshift.com/container-platform/4.15/virt/storage/virt-automatic-bootsource-updates.html

AFAIK, [this approach](https://github.com/openshift/openshift-docs/pull/74920#discussion_r1594179450) will work without any issues. Please note that the yaml file we are suggesting to edit is the configuration of the deployed components in charge of managing OCP-Virt. Therefore, a change to this configuration file will cause the underlying infrastructure to reconcile and pick up the updated configuration, i.e., enabling/disabling the feature gate.

Regarding the `HyperConverged CR` term definition, other docs entries, e.g., [CPU models](https://docs.openshift.com/container-platform/4.15/virt/nodes/virt-managing-node-labeling-obsolete-cpu-models.html), [FQDN](https://docs.openshift.com/container-platform/4.15/virt/vm_networking/virt-accessing-vm-secondary-network-fqdn.html) or [boot sources](https://docs.openshift.com/container-platform/4.15/virt/storage/virt-automatic-bootsource-updates.html), does not define the term HyperConverged CR, at most the first occurrence of `CR` is defined as `custom resource`. IMHO, the doc entries assume an educated reader in the basic terms.

Comment (geetikakay) [reply modules/virt-configuring-downward-metrics.adoc:15]: Enabling Downward metrics feature gate

1. Enable the downward metrics. Open the HyperConverged CR in your default editor by running the following command
$ oc edit hyperconverged kubevirt-hyperconverged -n openshift-cnv


Step 2: Edit the HyperConverged CR, to enable feature gate 


apiVersion: hco.kubevirt.io/v1beta1
kind: HyperConverged
metadata:
  name: kubevirt-hyperconverged
  namespace: openshift-cnv
spec:
  featureGates:
    downwardMetrics: false
# ...

Alternatively, To enable the downwardMetrics feature gate using the command line interface (CLI), use the oc patch command as follows:

$ oc patch hco kubevirt-hyperconverged -n openshift-cnv \
  --type json -p '[{"op": "replace", "path": \
  "/spec/featureGates/downwardMetrics" \
  "value": true}]'


Disabling Downward metrics feature gate
...

Just a thought , probably like this @aspauldi @jcanocan .

**Comment by aspauldi** [comment]
@jcanocan @geetikakay , I think (hope!) that I've gotten through all the suggestions you've made. Please review and let me know if you approve or if you feel changes are still needed. Thanks.

File: modules/virt-enabling-disabling-downward-metrics-feature gate-yaml.adoc [hunk 16]
```diff
@@ -0,0 +1,55 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-enabling-disabling-downward-metrics-feature-gate-yaml_{context}"]
+= Enabling and disabling the downward metrics feature gate in a YAML file
+
+To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by creating or editing a YAML file. Enabling the feature gate requires administrator privileges.
```
Comment (jcanocan) [inline modules/virt-enabling-disabling-downward-metrics-feature gate-yaml.adoc:9]:
```suggestion
To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by editing a YAML file. Enabling the feature gate requires administrator privileges.
```

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc [hunk 17]
```diff
@@ -0,0 +1,51 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-enabling-disabling-downward-metrics-feature-gate-yaml_{context}"]
+= Enabling or disabling the downward metrics feature gate in a YAML file
+
+To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by creating or editing a YAML file. Enabling the feature gate requires administrator privileges.
```
Comment (jcanocan) [inline modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc:9]:
```suggestion
To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by editing a YAML file. Enabling the feature gate requires administrator privileges.
```

File: modules/virt-viewing-downward-metrics-cli.adoc (hunk 15 above)
Comment (jcanocan) [reply modules/virt-viewing-downward-metrics-cli.adoc:17]: Still valid.

**Review by jcanocan — COMMENTED** [review]
Great job! Thanks for adding all the feedback provided. Just added a couple of suggestions.

File: modules/virt-viewing-downward-metrics-cli.adoc (hunk 15 above)
Comment (aspauldi) [reply modules/virt-viewing-downward-metrics-cli.adoc:17]: Sorry! Missed this last time.

**Comment by aspauldi** [comment]
@jcanocan and @geetikakay , can you please indicate your approval of the content? Thanks again for all your input.

**Comment by geetikakay** [comment]
lgtm

**Comment by jcanocan** [comment]
Thanks a lot once again for your tireless efforts! This is a really nice piece of documentation.

/lgtm

**Comment by skopacz1** [comment]
Accidentally added the done label, please ignore that for now 😅🤦‍♂️

File: modules/virt-about-downward-metrics.adoc [hunk 18]
```diff
@@ -0,0 +1,8 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-using-downward-metrics.adoc
```
Comment (skopacz1) [inline modules/virt-about-downward-metrics.adoc:3]: Nit - I can't see where this module is located. It's fine to keep this module in this PR if you're not using it, but if it's not in any assemblies, could you delete this one line so it doesn't seem like it's included anywhere?

```suggestion
```

File: modules/virt-configuring-downward-metrics.adoc [hunk 19]
```diff
@@ -0,0 +1,67 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-configuring-downward-metrics_virt-using-downward-metrics"]
```
Comment (skopacz1) [inline modules/virt-configuring-downward-metrics.adoc:6]: Per our [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#module-file-metadata), module IDs need a context at the end of them:
```suggestion
[id="virt-configuring-downward-metrics_virt-using-downward-metrics_{context}"]
```

File: modules/virt-disabling-downward-metrics-feature-gate.adoc
Comment (skopacz1) [inline modules/virt-disabling-downward-metrics-feature-gate.adoc:1]: I don't think I see this module used in the assembly. In this case, since there's already a procedure that includes this step, I would suggest deleting this file and removing it from your PR, but it's up to you. 

I'll still leave comments as if you're going to keep it.

File: modules/virt-disabling-downward-metrics-feature-gate.adoc [hunk 20]
```diff
@@ -0,0 +1,21 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
```
Comment (skopacz1) [inline modules/virt-disabling-downward-metrics-feature-gate.adoc:3]: I would delete this line if the module isn't going to be included anywhere:
```suggestion
```

File: modules/virt-disabling-downward-metrics-feature-gate.adoc [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-disabling-downward-metrics-feature-gate"]
```
Comment (skopacz1) [inline modules/virt-disabling-downward-metrics-feature-gate.adoc:6]: Same comment as above about module IDs:
```suggestion
[id="virt-disabling-downward-metrics-feature-gate_{context}"]
```

File: modules/virt-enabling-disabling-downward-metrics-feature gate-yaml.adoc
Comment (skopacz1) [inline modules/virt-enabling-disabling-downward-metrics-feature gate-yaml.adoc:1]: This module seems to be a near-duplicate of another module you use in the assembly. I would suggest deleting this module and removing it from your PR if you're not going to use it.

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-cli.adoc [hunk 22]
```diff
@@ -0,0 +1,33 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-enabling-disabling-downward-metrics-feature-gate-cli__{context}"]
```
Comment (skopacz1) [inline modules/virt-enabling-disabling-downward-metrics-feature-gate-cli.adoc:6]: I think there's an extra underscore here that can be deleted:
```suggestion
[id="virt-enabling-disabling-downward-metrics-feature-gate-cli_{context}"]
```

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-cli.adoc [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+= Enabling or disabling the downward metrics feature gate from the command line
+
+To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by using the command line. Enabling the feature gate requires administrator privileges.
```
Comment (skopacz1) [inline modules/virt-enabling-disabling-downward-metrics-feature-gate-cli.adoc:9]: I would suggest making the last sentence a prereq:
```suggestion
To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by using the command line. 

.Prerequisites

* You have access to the cluster with administrator privileges.
```

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc [hunk 24]
```diff
@@ -0,0 +1,53 @@
+// Module included in the following assemblies:
+//
+// * virt/monitoring/virt-exposing-downward-metrics.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-enabling-disabling-downward-metrics-feature-gate-yaml_{context}"]
+= Enabling or disabling the downward metrics feature gate in a YAML file
+
+To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by editing a YAML file. Enabling the feature gate requires administrator privileges.
```
Comment (skopacz1) [inline modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc:9]: Same as above about making the last sentence a prereq:
```suggestion
To expose downward metrics for a host virtual machine, you can enable the `downwardMetrics` feature gate by editing a YAML file. 

.Prerequisites

* You have access to the cluster with administrator privileges.
```

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+
+.Procedure
+
+You can enable or disable the `downwardMetrics` feature gate as follows:
```
Comment (skopacz1) [inline modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc:13]: I don't think we typically introduce procedures with a sentence like this once the `.Procedure` section has started. I get that you're trying to be consistent with the other procedure, but I think this sentence's meaning is already implied and it can probably be deleted:
```suggestion
```

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+
+. Open the HyperConverged custom resource (CM) in your default editor by running the following command:
```
Comment (skopacz1) [inline modules/virt-enabling-disabling-downward-metrics-feature-gate-yaml.adoc:15]: Typo:
```suggestion
. Open the HyperConverged custom resource (CR) in your default editor by running the following command:
```

File: modules/virt-enabling-downward-metrics-feature-gate.adoc
Comment (skopacz1) [inline modules/virt-enabling-downward-metrics-feature-gate.adoc:1]: I think this is another unused duplicate module, I would suggest removing this file

File: modules/virt-viewing-downward-metrics-cli.adoc [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-viewing-downward-metrics-cli_{context}"]
+= Viewing downward metrics by using the command line
+
+You can view downward metrics by entering a command from inside a guest virtual machine (VM).
+
+.Procedure
+
+* Run the following command:
```
Comment (skopacz1) [inline modules/virt-viewing-downward-metrics-cli.adoc:13]: Nit - there are technically two commands users are instructed to use in this step:
```suggestion
* Run the following commands:
```

File: modules/virt-viewing-downward-metrics-cli.adoc [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
++
+[source,terminal]
+----
+$ sudo sh -c 'printf "GET /metrics/XML\n\n" > /dev/virtio-ports/org.github.vhostmd.1'
+$ sudo cat /dev/virtio-ports/org.github.vhostmd.1
+----
```
Comment (skopacz1) [inline modules/virt-viewing-downward-metrics-cli.adoc:19]: Per our [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#single-command-per-code-block), we should only have one command per code block in order for the copy/paste button on the website to work. You can still keep these two commands in the same step, but they just need to be in their own blocks:
```suggestion
[source,terminal]
----
$ sudo sh -c 'printf "GET /metrics/XML\n\n" > /dev/virtio-ports/org.github.vhostmd.1'
----
+
[source,terminal]
----
$ sudo cat /dev/virtio-ports/org.github.vhostmd.1
----
```

File: virt/monitoring/virt-exposing-downward-metrics.adoc [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+As an administrator, you can expose a limited set of host and virtual machine (VM) metrics to a guest VM by first enabling a `downwardMetrics` feature gate and then configuring a `downwardMetrics` device.
+
+Users can view the metrics results by using the command line or the `vm-dump-metrics tool`.
+
+[id="virt-enabling-disabling-feature-gate"]
+== Enabling or disabling the downwardMetrics feature gate
+
+You can enable or disable the `downwardMetrics` feature gate by performing either of the following actions:
+
+* Editing the HyperConverged custom resource (CM) in your default editor
```
Comment (skopacz1) [inline virt/monitoring/virt-exposing-downward-metrics.adoc:18]: Typo:

```suggestion
* Editing the HyperConverged custom resource (CR) in your default editor
```

**Review by skopacz1 — COMMENTED** [review]
Left some comments but otherwise looks good to me! Great work. 

One thing I noticed is that there are a number of extra or duplicate modules that don't seem to be used in the assembly you create. Unless you plan to use any of these later, I would strongly suggest deleting them from your PR. It'll prevent any maintenance headaches in the future if anyone wants to make changes to this content.

File: modules/virt-enabling-disabling-downward-metrics-feature-gate-cli.adoc (hunk 22 above)
Comment (skopacz1) [reply modules/virt-enabling-disabling-downward-metrics-feature-gate-cli.adoc:6]: This one item from my peer review still needs to be addressed.

**Review by skopacz1 — COMMENTED** [review]
Just one feedback item from my peer review that was missed, once it's addressed the PR will be good to merge! Thanks

**Comment by aspauldi** [comment]
@skopacz1 , All set now! Thanks.
